packages/flutter_tools/lib/src/base/logger.dart

```diff
@@ -192,98 +289,226 @@ class _LogMessage {
 
-  static const String _bold  = '\u001B[1m';
-  static const String _reset = '\u001B[0m';
-  static const String _clear = '\u001B[2J\u001B[H';
+  bool get isDumb {
+    return (cursorBack == null ||
+            cursorForward == null ||
+            cursorUp == null ||
+            cursorDown == null ||
+            clearEOL == null ||
+            clearScreen == null);
+  }
 
-  bool supportsColor;
+  // Function keys.
+  final String keyF1  = _tputGetSequence('kf1',  orElse: '\u001BOP');
+  final String keyF5  = _tputGetSequence('kf5',  orElse: '\u001B[15~');
+  final String keyF6  = _tputGetSequence('kf6',  orElse: '\u001B[17~');
+  final String keyF10 = _tputGetSequence('kf10', orElse: '\u001B[21~');
 
-  String writeBold(String str) => supportsColor ? '$_bold$str$_reset' : str;
+  // Back one character.
+  final String cursorBack = _tputGetSequence('cub1');
 
-  String clearScreen() => supportsColor ? _clear : '\n\n';
+  // Forward one character.
+  final String cursorForward = _tputGetSequence('cuf1');
 
-  set singleCharMode(bool value) {
-    stdin.lineMode = !value;
-  }
+  // Up one character.
+  final String cursorUp = _tputGetSequence('cuu1');
+
+  // Down one character.
+  final String cursorDown = _tputGetSequence('cud1');
+
+  // Clear to end of line.
+  final String clearEOL = _tputGetSequence('el');
+
+  // Clear screen and home cursor.
+  final String clearScreen = _tputGetSequence('clear', orElse: '\n\n');
+
+  // Enter bold text mode.
+  final String boldText = _tputGetSequence('bold', orElse: '');
+
+  // Exit text attributes.
+  final String resetText = _tputGetSequence('sgr0', orElse: '');
+
+  int get cols => stdout.terminalColumns;
+
+  bool supportsColor;
 
-  /// Return keystrokes from the console.
-  ///
-  /// Useful when the console is in [singleCharMode].
-  Stream<String> get onCharInput => stdin.transform(ASCII.decoder);
+  // Convenience method for bolding text.
+  String toBold(String str) => '$boldText$str$resetText';
 }
 
 class _AnsiStatus extends Status {
-  _AnsiStatus(this.message) {
+  _AnsiStatus(Logger logger, this.message) : super(logger) {
     stopwatch = new Stopwatch()..start();
-
-    stdout.write('${message.padRight(51)}     ');
-    stdout.write('${_progress[0]}');
-
-    timer = new Timer.periodic(new Duration(milliseconds: 100), _callback);
   }
 
-  static final List<String> _progress = <String>['-', r'\', '|', r'/', '-', r'\', '|', '/'];
+  static final List<String> _progress =
+      <String>['-', r'\', '|', r'/', '-', r'\', '|', '/'];
 
   final String message;
   Stopwatch stopwatch;
-  Timer timer;
-  int index = 1;
-  bool live = true;
+  String _currentLine;
+  bool _shown = true;
 
-  void _callback(Timer timer) {
-    stdout.write('\b${_progress[index]}');
-    index = ++index % _progress.length;
+  void _hide() {
+    if (!_shown) {
+      return;
+    }
+    _shown = false;
+    if (_currentLine != null) {
+      for (int i = 0; i < _currentLine.length; i++) {
```

Comment:
`i++` is suboptimal in some languages with some poor compilers. So it's generally best practice to avoid it, so that new programmers don't see it and start using it in places they don't know are in fact using those poor compilers. `++i` avoids the problem but looks ugly, so I prefer `i += 1`. The latter also has the advantage of being easy to adjust to different increments in general.

Personally if it was up to me we'd get rid of the pre- and post- increment and decrement unary operators. They don't add anything useful to the language IMHO.

This isn't mentioned in the flutter style guide but we maybe should add it. I don't believe we use post-increment `++` anywhere in the code base, though some might have slipped through.